apps/o2-blocks/src/callout/editor.js

```diff
@@ -0,0 +1,331 @@
+				{ __( 'Selected:' ) } { currentIconName }
+			</div>
+		</div>
+	);
+};
+
+const TemplateControls = ( { templateName, onChange } ) => {
+	const templateOptions = useMemo(
+		() =>
+			Object.entries( CALLOUT_TEMPLATES ).map( ( [ key, value ] ) => ( {
+				key,
+				name: value.name,
+			} ) ),
+		[]
+	);
+
+	const selectedTemplate = useMemo(
+		() =>
+			templateOptions.find( ( option ) => option.key === templateName ) || {
+				key: 'custom',
+				name: CALLOUT_TEMPLATES.custom.name,
+			},
+		[ templateOptions, templateName ]
+	);
+
+	return (
+		<PanelBody title={ __( 'Templates' ) }>
+			<CustomSelectControl
+				options={ templateOptions }
+				value={ selectedTemplate }
+				onChange={ onChange }
+			/>
+		</PanelBody>
+	);
+};
+
+const ColorControls = ( { customColor, onColorChange } ) => (
+	<PanelBody title={ __( 'Color' ) }>
+		<ColorPalette
+			colors={ CALLOUT_PANEL_COLORS }
+			value={ customColor }
+			onChange={ onColorChange }
+		/>
+	</PanelBody>
+);
+
+const IconControls = ( { customIcon, currentColor, templateName, onIconChange } ) => (
+	<PanelBody title={ __( 'Icon' ) }>
+		<IconPicker
+			customIcon={ customIcon }
+			currentColor={ currentColor }
+			templateName={ templateName }
+			onIconChange={ onIconChange }
+		/>
+	</PanelBody>
+);
+
+// === MAIN COMPONENT ===
+
+const Edit = ( { attributes, setAttributes } ) => {
+	const {
+		calloutTemplate: templateName = 'custom',
+		customColor = CALLOUT_TEMPLATE_DEFAULT_COLOR,
+		customIcon = '',
+		headerContent = '',
+	} = attributes;
+
+	const currentColor = getCurrentColor( templateName, customColor );
+	const currentIcon = getIcon( templateName, customIcon );
+
+	const customStyle =
+		templateName === 'custom'
+			? { '--custom-color': normalizeColor( customColor ) || CALLOUT_TEMPLATE_DEFAULT_COLOR }
+			: {};
+
+	const handleTemplateChange = useCallback(
+		( { selectedItem } ) => {
+			if ( selectedItem?.key && selectedItem.key !== templateName ) {
+				const newAttributes = { calloutTemplate: selectedItem.key };
+
+				// Clear custom icon when switching to a template type (non-custom)
+				if ( selectedItem.key !== 'custom' && customIcon ) {
+					newAttributes.customIcon = '';
+				}
+
+				setAttributes( newAttributes );
+			}
+		},
+		[ templateName, customIcon, setAttributes ]
+	);
+
+	const handleColorChange = useCallback(
+		( color ) => {
+			const normalizedColor = normalizeColor( color );
+			if ( normalizedColor !== customColor ) {
+				setAttributes( { customColor: normalizedColor } );
+			}
+		},
+		[ customColor, setAttributes ]
+	);
+
+	const handleIconChange = useCallback(
+		( iconName ) => {
+			if ( iconName !== customIcon ) {
+				setAttributes( { customIcon: iconName } );
+			}
+		},
+		[ customIcon, setAttributes ]
+	);
+
+	const handleHeaderChange = useCallback(
+		( newHeaderContent ) => {
+			setAttributes( { headerContent: newHeaderContent } );
+		},
+		[ setAttributes ]
+	);
```

Comment:
for as simple as all of these are it would be helpful to avoid creating the callbacks. they can be inlined when setting the handlers.

also it should be fine to call `setAttributes()` even when the value isn’t changed, as the editor can already handle non-update updates.

Reply:
Makes sense! Making these handlers inline given how simple they are!